File: src/main/java/com/example/lee/firstRound/BinaryTreeUpsideDown.java
package com.example.lee.firstRound;

import com.example.lee.model.TreeNode;

/**
 * Created by benbendaisy on 3/16/15.
 *
 * Given a binary tree where all the right nodes are either leaf nodes with a sibling (a left node that shares the same parent node) or empty, flip it upside down and turn it into a tree where the original right nodes turned into left leaf nodes. Return the new root.
 * For example:
 * Given a binary tree {1,2,3,4,5},
 *     1
 *    / \
 *   2   3
 *  / \
 * 4   5
 * return the root of the binary tree [4,5,2,#,#,3,1].
 *      4
 *     / \
 *    5   2
 *       / \
 *      3   1
 */
public class BinaryTreeUpsideDown {

    //refer to http://www.cnblogs.com/EdwardLiu/p/4232896.html
    private TreeNode head = null;
    public TreeNode UpsideDownBinaryTree(TreeNode root) {
        if (null == root) {
            return null;
        }
        return head;
    }

    private TreeNode treeHelper(TreeNode root) {
        if (null == root.left) {
            head = root;
            return root;
        }
        TreeNode newRoot = treeHelper(root.left);
        newRoot.left = root.right;
        newRoot.right = root;
        return newRoot.right;
    }
}

File: python_module/examples/binary_tree_upside_down.py
# Definition for a binary tree node.
from typing import Optional


class TreeNode:
    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

class Solution:
    def upsideDownBinaryTree(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root or (not root.left and not root.right):
            return root

        node, parent, parentRight = root, None, None
        while node:
            left = node.left
            node.left = parentRight
            parentRight = node.right
            node.right = parent
            parent = node
            node = left

        return parent

    def upsideDownBinaryTree1(self, root: Optional[TreeNode]) -> Optional[TreeNode]:
        if not root or (not root.left and not root.right):
            return root

        new_root = self.upsideDownBinaryTree(root.left)
        left = root.left
        right = root.right
        left.left = right
        left.right = root
        root.left = root.right = None
        return new_root